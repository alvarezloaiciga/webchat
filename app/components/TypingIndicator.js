// @flow
import React from 'react';
import {supportsSVG} from 'utils/utils';
import './styles/TypingIndicator.scss';

export type TypingIndicatorProps = {
  color?: 'green' | 'blue',
  duration?: number,
  yScale?: number,
  xScale?: number,
};

const TypingIndicator = ({
  color = 'green',
  duration = 1,
  xScale = 1,
  yScale = 1,
}: TypingIndicatorProps) => {
  const getColor = () => {
    switch (color) {
      case 'blue':
        return '#a6daec';
      case 'green':
      default:
        return '#59ad5d';
    }
  };
  const width = xScale * 24;
  const height = yScale * 30;

  const opacityLayer = (begin: string) =>
    <animate
      attributeName="opacity"
      attributeType="XML"
      values="0.2; 1; .2"
      begin={begin}
      dur={`${duration}s`}
      repeatCount="indefinite"
    />;

  const heightLayer = (begin: string) => {
    const yStart = 10 * yScale;
    const yEnd = 20 * yScale;

    return (
      <animate
        attributeName="height"
        attributeType="XML"
        values={`${yStart}; ${yEnd}; ${yStart}`}
        begin={begin}
        dur={`${duration}s`}
        repeatCount="indefinite"
      />
    );
  };

  const yLayer = (begin: string) => {
    const yStart = 10 * yScale;
    const yEnd = 5 * yScale;

    return (
      <animate
        attributeName="y"
        attributeType="XML"
        values={`${yStart}; ${yEnd}; ${yStart}`}
        begin={begin}
        dur={`${duration}s`}
        repeatCount="indefinite"
      />
    );
  };

  const rect = (x: number, begin: string) => {
    const rectHeight = 10 * yScale;
    const rectWidth = 4 * xScale;

    return (
      <rect
        x={x * xScale}
        y={rectHeight}
        width={rectWidth}
        height={rectHeight}
        fill={getColor()}
        opacity="0.2"
      >
        {opacityLayer(begin)}
        {heightLayer(begin)}
        {yLayer(begin)}
      </rect>
    );
  };

  return supportsSVG()
    ? <svg
        className="TypingIndicator"
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        x="0"
        y="0"
        width={`${width}px`}
        height={`${height}px`}
        viewBox={`0 0 ${width} ${height}`}
        xmlSpace="preserve"
      >
        ...
        {rect(0, `${0 * duration}s`)}
        {rect(8, `${0.3 * duration}s`)}
        {rect(16, `${0.6 * duration}s`)}
      </svg>
    : <span className="plainText">...</span>;
};

export default TypingIndicator;
