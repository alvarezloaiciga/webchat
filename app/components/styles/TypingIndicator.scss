.TypingIndicator {
  width: 24px;
}
