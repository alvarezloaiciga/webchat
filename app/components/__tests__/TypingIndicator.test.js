// @flow
jest.mock('Common/Utils');
import React from 'react';
import type {TypingIndicatorProps} from '../TypingIndicator';
import TypingIndicator from '../TypingIndicator';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';

describe('TypingIndicator component', () => {
  let wrapper: ShallowWrapper;
  let testProps: TypingIndicatorProps;
  let render: () => void;

  beforeEach(() => {
    testProps = {
      color: 'blue',
      duration: 2,
      yScale: 2,
      xScale: 2,
    };
    render = () => {
      wrapper = shallow(<TypingIndicator {...testProps} />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders with minimal props', () => {
      testProps = {};
      render();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
