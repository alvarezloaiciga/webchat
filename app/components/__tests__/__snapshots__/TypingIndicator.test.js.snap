// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypingIndicator component rendering renders 1`] = `
<svg
  className="TypingIndicator"
  height="20px"
  style={Object {}}
  version="1.1"
  viewBox="0 0 50 20"
  width="50px"
  x="0"
  xmlSpace="preserve"
  xmlns="http://www.w3.org/2000/svg"
  y="0"
>
  <defs>
    <linearGradient
      gradientUnits="userSpaceOnUse"
      id="flash"
      x1="-300%"
      x2="-100%"
      y2="0%"
    >
      <stop
        offset="0%"
        stopColor="#333"
      />
      <stop
        offset="25%"
        stopColor="#ddd"
      />
      <stop
        offset="100%"
        stopColor="#333"
      />
      <animate
        attributeName="x1"
        dur="1.5s"
        from="-300%"
        repeatCount="indefinite"
        to="100%"
      />
      <animate
        attributeName="x2"
        dur="1.5s"
        from="-100%"
        repeatCount="indefinite"
        to="300%"
      />
    </linearGradient>
  </defs>
  <g>
    <title>
      Title!
    </title>
    <g
      fill="url(#flash)"
      id="background"
    >
      <rect
        height={20}
        width={25}
        x={12.5}
      />
      <circle
        cx={12.5}
        cy={10}
        r={10}
      />
      <circle
        cx={37.5}
        cy={10}
        r={10}
      />
    </g>
    <g
      id="foreground"
    >
      <g>
        <circle
          cx={50}
          cy={10}
          fill="white"
          r={3.3333333333333335}
          style={Object {}}
        >
          <animate
            attributeName="r"
            begin="0;o1.end"
            dur=".6"
            from={2}
            id="o0"
            to={4.666666666666667}
          />
        </circle>
        <circle
          cx={50}
          cy={10}
          fill="white"
          r={3.3333333333333335}
          style={Object {}}
        >
          <animate
            attributeName="r"
            begin="o0.end"
            dur=".6"
            from={4.666666666666667}
            id="o1"
            to={2}
          />
        </circle>
      </g>
      <g>
        <circle
          cx={100}
          cy={10}
          fill="white"
          r={3.3333333333333335}
          style={Object {}}
        >
          <animate
            attributeName="r"
            begin="1;o6.end"
            dur=".6"
            from={2}
            id="o5"
            to={4.666666666666667}
          />
        </circle>
        <circle
          cx={100}
          cy={10}
          fill="white"
          r={3.3333333333333335}
          style={Object {}}
        >
          <animate
            attributeName="r"
            begin="o5.end"
            dur=".6"
            from={4.666666666666667}
            id="o6"
            to={2}
          />
        </circle>
      </g>
      <g>
        <circle
          cx={150}
          cy={10}
          fill="white"
          r={3.3333333333333335}
          style={Object {}}
        >
          <animate
            attributeName="r"
            begin="2;o11.end"
            dur=".6"
            from={2}
            id="o10"
            to={4.666666666666667}
          />
        </circle>
        <circle
          cx={150}
          cy={10}
          fill="white"
          r={3.3333333333333335}
          style={Object {}}
        >
          <animate
            attributeName="r"
            begin="o10.end"
            dur=".6"
            from={4.666666666666667}
            id="o11"
            to={2}
          />
        </circle>
      </g>
    </g>
  </g>
</svg>
`;
